GetListOfEntriesOfLedger implementation

reddycharan opened the pull request:
Descriptions of the changes in this PR:

As described in this BP - https://github.com/apache/bookkeeper/blob/master/site/bps/BP-34-cluster-metadata-checker.md, this request returns list of entries bookie contains for the given ledger in an encoded format. The returned list provides weakly consistent state, of the entries of the ledger.

reddycharan commented on bookkeeper-proto/src/main/proto/BookkeeperProtocol.proto:
```diff
@@ -64,6 +64,7 @@ enum OperationType {
     GET_BOOKIE_INFO = 8;
     START_TLS = 9;
     FORCE_LEDGER = 10;
+    GET_LIST_OF_ENTRIES_OF_A_LEDGER = 11;
```
will change this to GET_LIST_OF_ENTRIES_OF_LEDGER and all the corresponding names will be changed to GetListOfEntriesOfLedger

reddycharan replied:
fixed it

reddycharan replied:
already fixed it

reddycharan commented on bookkeeper-server/src/main/java/org/apache/bookkeeper/bookie/Bookie.java:
```diff
@@ -1635,4 +1635,8 @@ static LedgerDirsManager createIndexDirsManager(ServerConfiguration conf, DiskCh
     }
+
+    public ByteBuf getListOfEntriesOfALedger(long ledgerId) throws IOException, NoLedgerException {
+        return null;
```
have to add corresponding method to LedgerDescriptor/LedgerDescriptorImpl and call that method here.

reddycharan replied:
fixed it

eolivelli commented on bookkeeper-server/src/main/java/org/apache/bookkeeper/bookie/storage/ldb/DbLedgerStorage.java:
```diff
@@ -335,4 +335,10 @@ public boolean isInForceGC() {
+    @Override
+    public byte[] getEntriesOfLedger(long ledgerId) {
+        // TODO Auto-generated method stub
+        return null;
```
not sure how can we deal with this

I guess we have to implement this one as well, or have some kind of protocol in which the bookie can report to the client something like 'I don't know which are the entries I have'

athanatos replied:
An error code would suffice -- getEntriesOfLedger should therefore be able to raise a NotImplemented exception or whatever is most idiomatic.

sijie replied:
I would suggest throw UnsupportedOperationException for now.

reddycharan replied:
fixed it

eolivelli commented on bookkeeper-server/src/main/java/org/apache/bookkeeper/proto/GetListOfEntriesOfALedgerProcessorV3.java:
```diff
@@ -0,0 +1,113 @@
+            LOG.error("No ledger found while performing getListOfEntriesOfALedger from ledger: {}", ledgerId, e);
+        } catch (IOException e) {
+            status = StatusCode.EIO;
+            LOG.error("IOException while performing getListOfEntriesOfALedger from ledger: {}", ledgerId);
```
any reason for not reporting the stacktrace of the exception
IOException are very generic and having the detail will ease debug

reddycharan replied:
will fix it

eolivelli commented on bookkeeper-server/src/main/java/org/apache/bookkeeper/bookie/InterleavedLedgerStorage.java:
```diff
@@ -641,4 +649,233 @@ public void onRotateEntryLog() {
+     * needed, so each SequenceGroup can be represented with (2 * 8 + 2 * 4 = 24
+     * bytes).
+     */
+    private static class SequenceGroup {
```
we should have specific unit tests around these classes, they can be tested without starting a bookie

reddycharan replied:
moved logic to AvailabilityOfEntriesOfLedger.java class and added relevant tests for this class

athanatos commented on bookkeeper-server/src/main/java/org/apache/bookkeeper/bookie/EntryMemTable.java:
```diff
@@ -456,4 +459,95 @@ boolean isEmpty() {
+            public EntryKey next() {
+                if (hasNext()) {
+                    EntryKey returnEntryKey = null;
+                    if (curKVMapEntry != null && curSnapshotEntry != null) {
```
Agreed.

reddycharan replied:
moved merging of iterator logic to bookkeeper-server/src/main/java/org/apache/bookkeeper/util/IteratorUtility.java

athanatos commented on bookkeeper-server/src/main/java/org/apache/bookkeeper/bookie/InterleavedLedgerStorage.java:
```diff
@@ -641,4 +649,233 @@ public void onRotateEntryLog() {
+        }
+    }
+
+    protected byte[] getEntriesOfLedger(long ledgerId, Iterator<EntryKey> entriesInMemtable) throws Exception {
```
This shouldn't take entriesInMemtable as an input -- InterleavedLedgerStorage doesn't necessarily require a memtable.  Separate the above representation classes out into their own files with unit tests.  Modify this method to return an Iterator.  SortedMemTable would be responsible for merging this iterator with the one returned from the memtable.  The SequenceGroup above could simply take an interator in its constructor.

reddycharan replied:
new logic should take care of this

athanatos commented on bookkeeper-server/src/main/java/org/apache/bookkeeper/bookie/InterleavedLedgerStorage.java:
```diff
@@ -641,4 +649,233 @@ public void onRotateEntryLog() {
+     * SequenceGroups will be appended sequentially to this byte array, with
+     * each SequenceGroup taking 24 bytes.
+     */
+    static class AvailabilityOfEntriesOfLedger {
```
Call this something more general -- LedgerEntrySet or something.

athanatos commented on bookkeeper-server/src/main/java/org/apache/bookkeeper/bookie/SortedLedgerStorage.java:
```diff
@@ -354,4 +355,10 @@ public boolean isInForceGC() {
     }
+
+    @Override
+    public byte[] getEntriesOfLedger(long ledgerId) throws Exception {
```
See: this shouldn't return a byte[] -- it should return an iterator representing the merged streams from the memtable and interleavedLedgerStorage.  The caller should be responsible for encoding it before it gets sent back.

reddycharan replied:
as mentioned in other comment - 

https://github.com/apache/bookkeeper/pull/1931/files/a9d00a25d29bd034c7ee070f7a0d75bac29a94d2#r254091827

reddycharan replied:
new logic should take care of this

athanatos commented on bookkeeper-server/src/main/java/org/apache/bookkeeper/proto/GetListOfEntriesOfALedgerProcessorV3.java:
```diff
@@ -0,0 +1,113 @@
+        StatusCode status = StatusCode.EOK;
+        ByteBuf listOfEntriesOfALedger = null;
+        try {
+            listOfEntriesOfALedger = requestProcessor.bookie.getListOfEntriesOfALedger(ledgerId);
```
The encoding should happen here.

eolivelli replied:
totally agree ! good idea @athanatos

eolivelli replied:
I am not really sure getListOfEntriesOfALedger should return an iterator.
what about leaving the scan inside bookie or the LedgerStorage impl

getListOfEntriesOfALedger(long ledgerId, Consumer<Long> entryId)

this will ease handling of errors and it may help LedgerStorage implementations which already have that piece of information

athanatos replied:
An Iterator based interface is likely to be easier for using a generic method for combining streams.

reddycharan replied:
@athanatos @eolivelli I don't think it is valid to do here for the following reasons -

1) getListOfEntriesOfALedger implementation is LedgerStorage specific, I don't see how can we abstract this out at LedgerStorage specific level and deal with it here. I don’t think it is optimal to expect all the Storage implementation to return Iterator of long values (entry numbers), because in each storage implementation, things might be different for how entry details can be read. Having iterator based approach is not ideal in all cases.
2) From what I understand, GetListOfEntriesOfALedgerProcessorV3 is a request/response processor and it is not appropriate to deal with business logic in this class.
3) Also this class is in proto package, we don’t want the Encoding class details / code to be in this package and also this class won’t have access to ‘bookie’ package protected classes.

athanatos replied:
For 2. I'd argue that it's not business logic, it's logic specific to the particular binary format you are using to ship it to the client.

For 1. I'm mainly arguing that you need to merge streams of sorted entries in at least two places, I'd like it to be uniform and an iterator seems the natural way to do that.  A visitor could also work.

3. That's a good argument for moving the class representing the encoding up into the proto package.

athanatos replied:
Also, listing the entries of a ledger seems like it would be applicable to any ledger manager.

reddycharan replied:
new logic should take care of this

sijie commented on bookkeeper-server/src/main/java/org/apache/bookkeeper/proto/BookkeeperInternalCallbacks.java:
```diff
@@ -107,6 +107,13 @@
+     * A callback interface for GetListOfEntriesOfALedger command.
+     */
+    public interface GetListOfEntriesOfALedgerCallback {
+        void getListOfEntriesOfALedgerComplete(int rc, long ledgerId, ByteBuf buffer, Object ctx);
```
I would suggest trying to use `CompletableFuture` on newly introduced features.

reddycharan replied:
introducing CompletableFuture

sijie commented on bookkeeper-server/src/main/java/org/apache/bookkeeper/bookie/LedgerStorage.java:
```diff
@@ -248,4 +248,20 @@ default boolean isInForceGC() {
+     *         ledgerstorage.
+     * @throws Exception
+     */
+    byte[] getEntriesOfLedger(long ledgerId) throws Exception;
```
I would suggest returning a stream/iterator of entries. The encoding should be separated from LedgerStorage implementation.

reddycharan replied:
new logic should take care of this

reddycharan commented:
@sijie @eolivelli @jvrao this is good for final review.

eolivelli commented:
I left some minor comments, not addresses yet.
And CI is red

reddycharan commented on bookkeeper-proto/src/main/proto/BookkeeperProtocol.proto:
```diff
@@ -152,6 +154,10 @@ message GetBookieInfoRequest {
 }
 
+message GetListOfEntriesOfALedgerRequest {
+	required int64 ledgerId = 1;
```
will fix indentation issues.

reddycharan commented on bookkeeper-server/src/main/java/org/apache/bookkeeper/bookie/EntryMemTable.java:
```diff
@@ -44,7 +47,7 @@
 public class EntryMemTable implements AutoCloseable{
     private static Logger logger = LoggerFactory.getLogger(Journal.class);
-
+    static final long LONG_MAX_VALUE = Long.MAX_VALUE - 1;
```
hmm..I could've directly used 'Long.MAX_VALUE', but I'm worried if anywhere they are incrementing the entryid value (be it for comparison or something), thats why to be on safe side I'm creating this static variable.

I dont see any issue in naming it LONG_MAX_VALUE.

reddycharan replied:
removed this static long value

reddycharan commented on bookkeeper-server/src/main/java/org/apache/bookkeeper/bookie/InterleavedLedgerStorage.java:
```diff
@@ -641,4 +649,233 @@ public void onRotateEntryLog() {
+        return getEntriesOfLedger(ledgerId, Collections.emptyIterator());
+    }
+
+    public static final long INVALID_ENTRYID = -1;
```
moved this logic to AvailabilityOfEntriesOfLedger and yeah I should make it private.

jvrao commented on bookkeeper-proto/src/main/proto/BookkeeperProtocol.proto:
```diff
@@ -213,5 +220,11 @@ message GetBookieInfoResponse {
+message GetListOfEntriesOfLedgerResponse {
+    required StatusCode status = 1;
+    required int64 ledgerId = 2;
+    optional bytes availabilityOfEntriesOfLedger = 3; // treated as array of bits indicating availability of entry at that particular index location
```
I think comment is not accurate. Please update accordingly

reddycharan commented:
run integration tests

reddycharan commented:
run bookkeeper-server client tests

reddycharan commented:
rebuild java11

reddycharan commented:
run bookkeeper-server remaining tests

reddycharan commented:
run bookkeeper-server client tests

reddycharan commented:
rebuild java11

reddycharan commented:
run bookkeeper-server remaining tests

sijie commented on bookkeeper-server/src/main/java/org/apache/bookkeeper/bookie/InterleavedLedgerStorage.java:
```diff
@@ -643,4 +655,57 @@ public void onRotateEntryLog() {
+            pageEntriesIteratorNonFinal = Collections.emptyIterator();
+        }
+        final Iterator<LedgerCache.PageEntries> pageEntriesIterator = pageEntriesIteratorNonFinal;
+        return new OfLong() {
```
It might be worth moving this class to a separate class so it is easy to write tests for it.

reddycharan replied:
moved it to LedgerCache class

sijie commented on bookkeeper-server/src/main/java/org/apache/bookkeeper/bookie/InterleavedLedgerStorage.java:
```diff
@@ -643,4 +655,57 @@ public void onRotateEntryLog() {
+        }
+        final Iterator<LedgerCache.PageEntries> pageEntriesIterator = pageEntriesIteratorNonFinal;
+        return new OfLong() {
+            long[] entriesInCurrentLedgerPage = new long[entriesPerLedgerEntryPage];
```
This seems to be an unnecessary allocation. I am wondering why can't you just use the LEP.

reddycharan replied:
adding new method in LEP to return iterator

sijie commented on bookkeeper-server/src/main/java/org/apache/bookkeeper/bookie/storage/ldb/DbLedgerStorage.java:
```diff
@@ -352,4 +353,10 @@ static long getLongVariableOrDefault(ServerConfiguration conf, String keyName, l
     }
+
+    @Override
+    public OfLong getListOfEntriesOfLedger(long ledgerId) throws IOException {
```
If you don't implement this, please create a github issue and comment it here.

reddycharan replied:
created https://github.com/apache/bookkeeper/issues/2078 and adding comment in the code

reddycharan commented on bookkeeper-server/src/main/java/org/apache/bookkeeper/bookie/EntryMemTable.java:
```diff
@@ -456,4 +459,95 @@ boolean isEmpty() {
+                    } else if (curKVMapEntry != null) {
+                        returnEntryKey = curKVMapEntry;
+                        curKVMapEntry = null;
+                    } else {
```
i dont want to over complicate the utility iterator merger. I would leave a comment saying that IteratorUtility.mergePrimitiveLongIterator and IteratorUtility.mergeIteratorsForPrimitiveLongIterator expects input iterators to be sorted.

reddycharan commented:
@sijie @eolivelli addressed review comments and this should be good to go. Replication tests failure isn't related to my change, not sure why they are flapping.

reddycharan commented:
run bookkeeper-server replication tests

reddycharan commented:
rebuild java11

reddycharan commented:
Thanks @sijie @eolivelli @athanatos @jvrao for the review. I'm merging this PR.
